code/modules/mob/living/basic/bots/floorbot/floorbot.dm

```diff
@@ -0,0 +1,256 @@
+			for(var/a in 1 to iterations)
+				if(a == iterations)
+					tilestack.split_stack(null, tilestack.amount - tilestack.max_amount)
+				else
+					tilestack.split_stack(null, tilestack.max_amount)
+		tilestack = null
+
+
+/mob/living/basic/bot/floorbot/reset_bot(caller, turf/waypoint, message)
+	. = ..()
+	toggle_magnet(FALSE)
+
+/mob/living/basic/bot/floorbot/attackby(obj/item/W , mob/user, params)
+	if(istype(W, /obj/item/stack/tile/iron))
+		to_chat(user, span_notice("The floorbot can produce normal tiles itself."))
+		return
+	if(istype(W, /obj/item/stack/tile))
+		var/old_amount = tilestack ? tilestack.amount : 0
+		var/obj/item/stack/tile/tiles = W
+		if(tilestack)
+			if(!tiles.can_merge(tilestack))
+				to_chat(user, span_warning("Different custom tiles are already inside the floorbot."))
+				return
+			if(tilestack.amount >= maxtiles)
+				to_chat(user, span_warning("The floorbot can't hold any more custom tiles."))
+				return
+			tiles.merge(tilestack, maxtiles)
+		else
+			if(tiles.amount > maxtiles)
+				tilestack = tilestack.split_stack(null, maxtiles)
+			else
+				tilestack = W
+			tilestack.forceMove(src)
+		to_chat(user, span_notice("You load [tilestack.amount - old_amount] tiles into the floorbot. It now contains [tilestack.amount] tiles."))
+		return
+	else
+		..()
+
+/mob/living/basic/bot/floorbot/emag_act(mob/user)
+	. = ..()
+	if(!(bot_cover_flags & BOT_COVER_EMAGGED))
+		return
+	if(user)
+		to_chat(user, span_danger("[src] buzzes and beeps."))
+
+///mobs should use move_resist instead of anchored.
+/mob/living/basic/bot/floorbot/proc/toggle_magnet(engage = TRUE, change_icon = TRUE)
+	if(engage)
+		ADD_TRAIT(src, TRAIT_IMMOBILIZED, BUSY_FLOORBOT_TRAIT)
+		move_resist = INFINITY
+		if(change_icon)
+			icon_state = "[toolbox_color]floorbot-c"
+	else
+		REMOVE_TRAIT(src, TRAIT_IMMOBILIZED, BUSY_FLOORBOT_TRAIT)
+		move_resist = initial(move_resist)
+		if(change_icon)
+			update_icon()
+
+// Variables sent to TGUI
+/mob/living/basic/bot/floorbot/ui_data(mob/user)
+	var/list/data = ..()
+	if(!(bot_cover_flags & BOT_COVER_LOCKED) || issilicon(user) || isAdminGhostAI(user))
+		data["custom_controls"]["tile_hull"] = (floorbot_mode_flags & FLOORBOT_AUTO_TILE)
+		data["custom_controls"]["place_tiles"] =  (floorbot_mode_flags & FLOORBOT_PLACE_TILES)
+		data["custom_controls"]["place_custom"] = (floorbot_mode_flags & FLOORBOT_REPLACE_TILES)
+		data["custom_controls"]["repair_damage"] = (floorbot_mode_flags & FLOORBOT_FIX_FLOORS)
+		data["custom_controls"]["traction_magnets"] = !!HAS_TRAIT_FROM(src, TRAIT_IMMOBILIZED, BUSY_FLOORBOT_TRAIT)
+		data["custom_controls"]["tile_stack"] = 0
+		data["custom_controls"]["line_mode"] = FALSE
+		if(tilestack)
+			data["custom_controls"]["tile_stack"] = tilestack.amount
+		if(targetdirection)
+			data["custom_controls"]["line_mode"] = dir2text(targetdirection)
+	return data
+
+// Actions received from TGUI
```

Comment:
No need to comment on an inherited proc like this.
Also, make sure you're returning TRUE on each successful ui action